File: 2023/day05/go/solution.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

// MapRange represents a single mapping range
type MapRange struct {
	dstStart int
	srcStart int
	length   int
}

// Range represents a range of values [start, end)
type Range struct {
	start int
	end   int
}

func parseInput(text string) ([]int, [][]MapRange) {
	sections := strings.Split(strings.TrimSpace(text), "\n\n")

	// Parse seeds
	seedParts := strings.Split(sections[0], ": ")
	seedStrs := strings.Fields(seedParts[1])
	seeds := make([]int, len(seedStrs))
	for i, s := range seedStrs {
		seeds[i], _ = strconv.Atoi(s)
	}

	// Parse maps
	maps := make([][]MapRange, len(sections)-1)
	for i, section := range sections[1:] {
		lines := strings.Split(strings.TrimSpace(section), "\n")
		ranges := make([]MapRange, len(lines)-1)
		for j, line := range lines[1:] { // Skip header
			parts := strings.Fields(line)
			dstStart, _ := strconv.Atoi(parts[0])
			srcStart, _ := strconv.Atoi(parts[1])
			length, _ := strconv.Atoi(parts[2])
			ranges[j] = MapRange{dstStart, srcStart, length}
		}
		maps[i] = ranges
	}

	return seeds, maps
}

func applyMap(value int, ranges []MapRange) int {
	for _, r := range ranges {
		if value >= r.srcStart && value < r.srcStart+r.length {
			return r.dstStart + (value - r.srcStart)
		}
	}
	return value
}

func seedToLocation(seed int, maps [][]MapRange) int {
	value := seed
	for _, mapRanges := range maps {
		value = applyMap(value, mapRanges)
	}
	return value
}

func part1(seeds []int, maps [][]MapRange) int {
	minLoc := seedToLocation(seeds[0], maps)
	for _, seed := range seeds[1:] {
		loc := seedToLocation(seed, maps)
		if loc < minLoc {
			minLoc = loc
		}
	}
	return minLoc
}

func applyMapToRanges(inputRanges []Range, mapRanges []MapRange) []Range {
	var result []Range

	for _, ir := range inputRanges {
		remaining := []Range{{ir.start, ir.end}}

		for _, mr := range mapRanges {
			srcEnd := mr.srcStart + mr.length
			var newRemaining []Range

			for _, r := range remaining {
				// Part before the map range (unmapped)
				if r.start < mr.srcStart {
					end := r.end
					if end > mr.srcStart {
						end = mr.srcStart
					}
					newRemaining = append(newRemaining, Range{r.start, end})
				}

				// Part within the map range (mapped)
				overlapStart := r.start
				if overlapStart < mr.srcStart {
					overlapStart = mr.srcStart
				}
				overlapEnd := r.end
				if overlapEnd > srcEnd {
					overlapEnd = srcEnd
				}
				if overlapStart < overlapEnd {
					offset := mr.dstStart - mr.srcStart
					result = append(result, Range{overlapStart + offset, overlapEnd + offset})
				}

				// Part after the map range (unmapped)
				if r.end > srcEnd {
					start := r.start
					if start < srcEnd {
						start = srcEnd
					}
					newRemaining = append(newRemaining, Range{start, r.end})
				}
			}

			remaining = newRemaining
		}

		// Any remaining parts are unmapped (identity)
		result = append(result, remaining...)
	}

	return result
}

func part2(seeds []int, maps [][]MapRange) int {
	// Convert seeds to ranges: pairs of (start, start + length)
	var ranges []Range
	for i := 0; i < len(seeds); i += 2 {
		start := seeds[i]
		length := seeds[i+1]
		ranges = append(ranges, Range{start, start + length})
	}

	// Apply each map to the ranges
	for _, mapRanges := range maps {
		ranges = applyMapToRanges(ranges, mapRanges)
	}

	// Find minimum start of any range
	minStart := ranges[0].start
	for _, r := range ranges[1:] {
		if r.start < minStart {
			minStart = r.start
		}
	}
	return minStart
}

func main() {
	data, err := os.ReadFile("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	seeds, maps := parseInput(string(data))

	fmt.Println("Part 1:", part1(seeds, maps))
	fmt.Println("Part 2:", part2(seeds, maps))
}

File: 2023/day05/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use File::Spec;

# Read input file
my $script_dir = dirname(__FILE__);
my $input_path = File::Spec->catfile($script_dir, '..', 'input.txt');
open my $fh, '<', $input_path or die "Cannot open $input_path: $!";
my $text = do { local $/; <$fh> };
close $fh;

# Parse input
my @sections = split /\n\n/, $text;

# Parse seeds
my ($seed_line) = $sections[0] =~ /seeds:\s+(.+)/;
my @seeds = split /\s+/, $seed_line;

# Parse maps
my @maps;
for my $i (1 .. $#sections) {
    my @lines = split /\n/, $sections[$i];
    shift @lines;  # Skip header
    my @ranges;
    for my $line (@lines) {
        next unless $line =~ /\S/;
        my @nums = split /\s+/, $line;
        push @ranges, [$nums[0], $nums[1], $nums[2]];  # dst_start, src_start, length
    }
    push @maps, \@ranges;
}

# Part 1: Apply maps to each seed and find minimum location
sub apply_map {
    my ($value, $ranges) = @_;
    for my $range (@$ranges) {
        my ($dst_start, $src_start, $length) = @$range;
        if ($value >= $src_start && $value < $src_start + $length) {
            return $dst_start + ($value - $src_start);
        }
    }
    return $value;
}

sub seed_to_location {
    my ($seed, $maps) = @_;
    my $value = $seed;
    for my $map_ranges (@$maps) {
        $value = apply_map($value, $map_ranges);
    }
    return $value;
}

sub part1 {
    my ($seeds, $maps) = @_;
    my $min_loc;
    for my $seed (@$seeds) {
        my $loc = seed_to_location($seed, $maps);
        $min_loc = $loc if !defined($min_loc) || $loc < $min_loc;
    }
    return $min_loc;
}

# Part 2: Work with ranges
sub apply_map_to_ranges {
    my ($input_ranges, $map_ranges) = @_;
    my @result;

    for my $input_range (@$input_ranges) {
        my ($start, $end) = @$input_range;
        my @remaining = ([$start, $end]);

        for my $map_range (@$map_ranges) {
            my ($dst_start, $src_start, $length) = @$map_range;
            my $src_end = $src_start + $length;
            my @new_remaining;

            for my $rem (@remaining) {
                my ($r_start, $r_end) = @$rem;

                # Part before the map range (unmapped)
                if ($r_start < $src_start) {
                    push @new_remaining, [$r_start, $r_end < $src_start ? $r_end : $src_start];
                }

                # Part within the map range (mapped)
                my $overlap_start = $r_start > $src_start ? $r_start : $src_start;
                my $overlap_end = $r_end < $src_end ? $r_end : $src_end;
                if ($overlap_start < $overlap_end) {
                    my $offset = $dst_start - $src_start;
                    push @result, [$overlap_start + $offset, $overlap_end + $offset];
                }

                # Part after the map range (unmapped)
                if ($r_end > $src_end) {
                    my $after_start = $r_start > $src_end ? $r_start : $src_end;
                    push @new_remaining, [$after_start, $r_end];
                }
            }

            @remaining = @new_remaining;
        }

        # Any remaining parts are unmapped (identity)
        push @result, @remaining;
    }

    return \@result;
}

sub part2 {
    my ($seeds, $maps) = @_;

    # Convert seeds to ranges: pairs of (start, start + length)
    my @ranges;
    for (my $i = 0; $i < @$seeds; $i += 2) {
        my $start = $seeds->[$i];
        my $length = $seeds->[$i + 1];
        push @ranges, [$start, $start + $length];
    }

    # Apply each map to the ranges
    my $current_ranges = \@ranges;
    for my $map_ranges (@$maps) {
        $current_ranges = apply_map_to_ranges($current_ranges, $map_ranges);
    }

    # Find minimum start of any range
    my $min_start;
    for my $range (@$current_ranges) {
        my $start = $range->[0];
        $min_start = $start if !defined($min_start) || $start < $min_start;
    }

    return $min_start;
}

# Run both parts
print "Part 1: ", part1(\@seeds, \@maps), "\n";
print "Part 2: ", part2(\@seeds, \@maps), "\n";
